Profile settings form

jurgenwerk opened the pull request:
(This is a continuation of work that @backspace did before leaving for vacation.)

We're adding a profile settings screen, where the first pass only includes the ability to edit the display name. Here you can see this in action:

https://github.com/cardstack/boxel/assets/273660/2daa0030-f3e1-4d84-b844-ad208cd67625

However, the ticket is mentioning we should consider auto saving but I feel our auto saving mechanisms have sometimes brought us more problems than not... And as @backspace already brought up [in the ticket](https://linear.app/cardstack/issue/CS-6142/introduce-settings-screen-with-ability-to-set-name#comment-b85394b7) - the auto save behaviour for this case is currently not well defined. That's why I think that having a save button is a much simpler solution for now which can be replaced with automatism later on if needed. 

I also improvised the save success indicator by having different states in the save button (check how "Saved!" gets displayed for a second–to let the user know saving worked):

https://github.com/cardstack/boxel/assets/273660/f40c796e-1b6e-4511-80dd-a2c70c38337f

Let me know if you have any comments regarding the saving behaviour?

TODO:
- [x] Add error display (if request to change the display name fails)
- [x] Basic validation (prevent empty strings?)

backspace committed:
Add preliminary settings button

backspace committed:
Add preliminary settings modal

backspace committed:
Add profile component in settings modal

backspace committed:
Fix assertions to not expect element

backspace committed:
Change to use ModalContainer

Should this be in boxel-ui?

backspace committed:
Add sidebar option to ModalContainer

backspace committed:
Merge from main

backspace committed:
Remove border from large user icon

backspace committed:
Add display name to profile component

backspace committed:
Add animation when opening account popover

This doesn’t animate when closing… seems hackish, should
we be using boxel-motion?

backspace committed:
Add listener for name change events

It works but the test fails!

backspace committed:
Remove unnecessary helper

backspace committed:
Add more logging

backspace committed:
Merge from main

jurgenwerk committed:
Merge branch 'main' into host/settings-cs-6142

jurgenwerk committed:
wip

jurgenwerk committed:
Refactor

jurgenwerk committed:
Merge branch 'main' into host/settings-cs-6142

jurgenwerk committed:
Refactor some more

jurgenwerk committed:
Test

jurgenwerk committed:
Reload profile on display name event

jurgenwerk committed:
Pull footer buttons to the right

jurgenwerk commented:
@cardstack/boxel-developers this can already be reviewed - I only have a couple non-major todos pending.

jurgenwerk committed:
Pull footer buttons to the right in the catalog chooser

lukemelia commented:
I intended this UI to be similar to Interact submode and potentially implemented with cards in the future. Therefore, I would like to have it be equivalent in terms of auto-save and to solve the problems that we have with it in both places. That said, it is OK to have the save button as iterative step and to spin off a separate task for auto-save

habdelra commented on packages/host/app/components/operator-mode/profile-settings-modal.gts:
```diff
@@ -0,0 +1,136 @@
+
+  private saveTask = restartableTask(async () => {
+    await this.matrixService.profile.loaded; // Prevent saving before profile is loaded
+    let delayMs = config.environment === 'test' ? 1 : 1000;
```
let's just put the delay in the config and have the delay for the test env be 0. Simliar to these: 
https://github.com/cardstack/boxel/blob/main/packages/host/config/environment.js#L73-L77

it's much simpler that way.

jurgenwerk replied:
Ah I see, I adjusted this.

habdelra commented on packages/matrix/tests/login.spec.ts:
we also need a test that actually changes the display name via the profile dialog too

jurgenwerk replied:
Added!

jurgenwerk committed:
Add basic validation and make buttons more consistent with the rest of the app

jurgenwerk committed:
Improve & add tests

jurgenwerk commented:
@lukemelia I see, thanks. I will take care of noting that.

I added a couple of improvements:

1. Validation message when entering a blank display name:

<img width="668" alt="image" src="https://github.com/cardstack/boxel/assets/273660/a9f4e647-fe14-4fd2-b00c-1468748e5044">

2. Generic error message when a request to update display name fails:

<img width="977" alt="image" src="https://github.com/cardstack/boxel/assets/273660/799241b4-9514-4960-b702-13460626b0d2">

3. Added a cancel button and buttons styles that are more consistent with the rest of the dialogs we have:

<img width="257" alt="image" src="https://github.com/cardstack/boxel/assets/273660/0791e75c-9237-4d2b-a286-a46926454e6c">


This is now ready for another review - even though I'm still [fixing the reactivity when display name changes from outside](https://github.com/cardstack/boxel/pull/936/files#diff-b278e089a693c763f09d8d01c9587aee5608bc486a51e0d1521099105d1673fdR82) (I temporarily marked that test as skipped). I realize this PR is blocking another settings related task so this can be merged while I'm addressing the mentioned issue.

jurgenwerk committed:
In profile avatar show the initials depending on whether display name is set

habdelra commented:
I'm not super concerned about the profile name changing from the outside. i think the way our matrix use is shaping up is that our host app will be the only client that drives the profile name change

jurgenwerk commented:
@habdelra That's good to know, in that case we can just skip this feature. I'll remove that handler.

habdelra commented:
I noticed a minor nit--when the profile name is not different in the input field from the current name, the save button is enabled. I think we can disable the save button if you haven't made a change to the name

habdelra commented:
@jurgenwerk I'll push this PR forward since you are on break--happy holidays!

habdelra commented on packages/matrix/tests/login.spec.ts:
```diff
@@ -73,6 +79,31 @@ test.describe('Login', () => {
+      'newname',
+    );
+  });
+
```
i think we can remove this test--seems like an unlikely scenario given the current requirements

habdelra commented on packages/host/app/services/matrix-service.ts:
```diff
@@ -100,9 +104,14 @@ export default class MatrixService extends Service {
         Membership.onMembership(this),
       ],
       [this.matrixSDK.RoomEvent.Timeline, Timeline.onTimeline(this)],
+      [this.matrixSDK.UserEvent.DisplayName, this.onDisplayNameChange],
```
probably we don't need this since profile name changes will only ever come from the host app of the currently logged in user...

habdelra committed:
disable profile save button if display name unchanged

habdelra committed:
remove display name change handler
